src/lib/blkalloc/bitmap_blk_allocator.h

```diff
@@ -0,0 +1,121 @@
+#include <thread>
+#include <vector>
+
+#include <sisl/fds/bitset.hpp>
+#include <folly/MPMCQueue.h>
+#include <sisl/utility/enum.hpp>
+#include <sisl/utility/urcu_helper.hpp>
+#include <sisl/fds/thread_vector.hpp>
+
+#include <homestore/homestore_decl.hpp>
+#include <homestore/blk.h>
+#include "common/homestore_config.hpp"
+#include "common/homestore_assert.hpp"
+
+#include "blk_allocator.h"
+
+namespace homestore {
+
+class BlkAllocPortion {
+private:
+    mutable std::mutex m_blk_lock;
+    blk_num_t m_portion_num;
+    blk_temp_t m_temperature;
+
+public:
+    BlkAllocPortion(blk_temp_t temp = default_temperature()) : m_temperature(temp) {}
+    ~BlkAllocPortion() = default;
+    BlkAllocPortion(BlkAllocPortion const&) = delete;
+    BlkAllocPortion(BlkAllocPortion&&) noexcept = delete;
+    BlkAllocPortion& operator=(BlkAllocPortion const&) = delete;
+    BlkAllocPortion& operator=(BlkAllocPortion&&) noexcept = delete;
+
+    auto portion_auto_lock() const { return std::scoped_lock< std::mutex >(m_blk_lock); }
+    blk_num_t get_portion_num() const { return m_portion_num; }
+    blk_temp_t temperature() const { return m_temperature; }
+
+    void set_portion_num(blk_num_t portion_num) { m_portion_num = portion_num; }
+    void set_temperature(const blk_temp_t temp) { m_temperature = temp; }
+    static constexpr blk_temp_t default_temperature() { return 1; }
+};
+
+class CP;
+class BitmapBlkAllocator : public BlkAllocator {
+public:
+    BitmapBlkAllocator(BlkAllocConfig const& cfg, bool is_fresh, chunk_num_t id = 0);
+    BitmapBlkAllocator(BlkAllocator const&) = delete;
+    BitmapBlkAllocator(BitmapBlkAllocator&&) noexcept = delete;
+    BitmapBlkAllocator& operator=(BitmapBlkAllocator const&) = delete;
+    BitmapBlkAllocator& operator=(BitmapBlkAllocator&&) noexcept = delete;
+    virtual ~BitmapBlkAllocator() = default;
+
+    virtual void load() = 0;
+    BlkAllocStatus alloc_on_disk(BlkId const& in_bid) override;
+    void free_on_disk(BlkId const& b) override;
+    bool is_blk_alloced_on_disk(BlkId const& b, bool use_lock = false) const override;
+    void cp_flush(CP* cp) override;
+
+    blk_num_t get_num_portions() const { return (m_num_blks - 1) / m_blks_per_portion + 1; }
+    blk_num_t get_blks_per_portion() const { return m_blks_per_portion; }
+
+    BlkAllocPortion& get_blk_portion(blk_num_t portion_num) {
+        HS_DBG_ASSERT_LT(portion_num, get_num_portions(), "Portion num is not in range");
+        return m_blk_portions[portion_num];
+    }
+
+    blk_num_t blknum_to_portion_num(const blk_num_t blknum) const { return blknum / m_blks_per_portion; }
+    BlkAllocPortion& blknum_to_portion(blk_num_t blknum) { return m_blk_portions[blknum_to_portion_num(blknum)]; }
+    BlkAllocPortion const& blknum_to_portion_const(blk_num_t blknum) const {
+        return m_blk_portions[blknum_to_portion_num(blknum)];
+    }
+
+    sisl::Bitset const* get_disk_bitmap() const { return is_persistent() ? m_disk_bm.get() : nullptr; }
```

Comment:
So the rule is if the api doesn't have xxx_const, it is by default is const, for mutable, we need xxx_mutable.

Reply:
I don't think there is any set rule, but new code we can try to do this.